### PR TITLE
Serverside hidden activities

### PR DESCRIPTION
**Pull request by raikasdev**
Very epic Rust code. Probably not very good looking and I will be sticking with TS from now.

Adds support for hiding activities for specific projects. This has been implemented client-side with hashes on the VSCode extension but I'd personally prefer to myself see the projects with names. Currently hidden projects can't be distinguished from each other and they will all have a empty string as the project name along with `"hidden": true`.

### CODE REVIEW

File: migrations/2024-03-12-174132_serverside_hidden/down.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+-- This file should undo anything in `up.sql`
+ALTER TABLE "coding_activities" DROP COLUMN "hidden";
+
```
Comment (lajp) [inline migrations/2024-03-12-174132_serverside_hidden/down.sql]: What are these empty lines?

File: migrations/2024-03-12-174132_serverside_hidden/up.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-- Your SQL goes here
+ALTER TABLE "coding_activities" ADD COLUMN "hidden" BOOL;
+UPDATE "coding_activities" SET "hidden" = false WHERE hidden IS NULL;
+ALTER TABLE "coding_activities" ALTER COLUMN hidden SET NOT NULL;
+
```
Comment (lajp) [inline migrations/2024-03-12-174132_serverside_hidden/up.sql]: what are these empty lines

File: src/database/activity.rs [hunk 3]
```diff
@@ -78,7 +81,19 @@ impl super::DatabaseWrapper {
...
 
         let mut conn = self.db.get().await?;
-        Ok(query.load::<CodingActivity>(&mut conn).await?)
+        let mut activities = query.load::<CodingActivity>(&mut conn).await?;
+
+        // Change hidden entries project name
+        if !is_self {
+            for act in &mut activities {
+                if act.hidden {
+                    // Empty string instead of None() to make sure we don't make everything into "undefined" :D
```
Comment (lajp) [inline src/database/activity.rs]: I'm not sure if we wan't to use an empty String here

File: src/database/activity.rs [hunk 4]
```diff
@@ -78,7 +81,19 @@ impl super::DatabaseWrapper {
         };
 
         let mut conn = self.db.get().await?;
-        Ok(query.load::<CodingActivity>(&mut conn).await?)
+        let mut activities = query.load::<CodingActivity>(&mut conn).await?;
+
+        // Change hidden entries project name
```
Comment (lajp) [inline src/database/activity.rs:86]: This is something that could be done in the query

File: src/api/friends.rs [hunk 5]
```diff
@@ -69,8 +72,11 @@ pub async fn get_friends(
             username: fwt.user.username,
             coding_time: fwt.coding_time,
             status: heartbeats.get(&fwt.user.id).map(|heartbeat| {
-                let (inner_heartbeat, start_time, duration) = heartbeat.to_owned();
+                let (mut inner_heartbeat, start_time, duration) = heartbeat.to_owned();
                 drop(heartbeat);
+                if inner_heartbeat.hidden == Some(true) {
+                    inner_heartbeat.project_name = Some("".to_string());
```
Comment (lajp) [inline src/api/friends.rs]: I'd prefer String::new()

File: src/api/friends.rs [hunk 6]
```diff
@@ -39,8 +39,11 @@ pub async fn add_friend(
                 username: friend.username.clone(),
                 coding_time: db.get_coding_time_steps(friend.id).await,
                 status: heartbeats.get(&friend.id).map(|heartbeat| {
-                    let (inner_heartbeat, start_time, duration) = heartbeat.to_owned();
+                    let (mut inner_heartbeat, start_time, duration) = heartbeat.to_owned();
                     drop(heartbeat);
+                    if inner_heartbeat.hidden == Some(true) {
+                        inner_heartbeat.project_name = Some("".to_string());
```
Comment (lajp) [inline src/api/friends.rs]: String::new() is better imo

File: src/api/users.rs [hunk 7]
```diff
@@ -115,11 +119,16 @@ pub async fn get_current_activity(
...
             } else {
-                let current_heartbeat = CurrentActivity {
+                let mut current_heartbeat = CurrentActivity {
                     started: start,
                     duration: duration.num_seconds(),
                     heartbeat: inner_heartbeat,
                 };
+
+                if !is_self && current_heartbeat.heartbeat.hidden == Some(true) {
+                    current_heartbeat.heartbeat.project_name = Some("".to_string());
```
Comment (lajp) [inline src/api/users.rs]: String::new()

**Review by lajp — CHANGES_REQUESTED** [review]
Hi, thanks for the PR!
This is definitely a functionality we want.
Here are some minor things I'd like you to address before we can merge this.

**Comment by lajp** [comment]
Additionally would it be possible for you or someone else (@Eldemarkki) to implement the funcitonality to the [Frontend side](https://github.com/Testaustime/testaustime-frontend) so that we can enable it immediately after merging?

File: migrations/2024-03-12-174132_serverside_hidden/up.sql (hunk 2 above)
Comment (raikasdev) [reply migrations/2024-03-12-174132_serverside_hidden/up.sql]: I don't know, I used diesel to autogenerate 🤷

Comment (lajp) [reply migrations/2024-03-12-174132_serverside_hidden/up.sql]: Ok, can you just remove them :point_right: :point_left:

File: src/database/activity.rs (hunk 4 above)
Comment (raikasdev) [reply src/database/activity.rs:86]: It could, but the pro-coder catgirl (Antti) thinks it's not worth trying to get the CASE statement with Diesel.

**Comment by lajp** [comment]
Should we also automatically hide new heartbeats for projects where there exists a hidden project with the same name and from the same user, regardless of what the client says? Or should we just leave the responsibility to the client?

**Review by lajp — APPROVED** [review]

**Review by lajp — CHANGES_REQUESTED** [review]
Please fix the tests, otherwise lgtm

**Review by DrVilepis — APPROVED** [review]
kabum

**Review by lajp — APPROVED** [review]
Looks good! Thanks for your contribution and sorry for the slow review.

**Pull request merged**